test/test_native_mha.py

```diff
@@ -0,0 +1,301 @@
+            if pad_all:
+                do_pad_all((ypt, ynpt))
+            # Zero the last row of each TxT weight matrix
+            if need_weights:
+                if average_attn_weights:
+                    weight_pt[0][-1] = torch.zeros_like(weight_pt[0][-1], device=device, dtype=dtype)
+                    weight_npt[0][-1] = torch.zeros_like(weight_npt[0][-1], device=device, dtype=dtype)
+                    if pad_all:
+                        do_pad_all((weight_pt, weight_npt))
+                else:
+                    for nh in range(num_heads):
+                        weight_pt[0][nh][-1] = torch.zeros_like(weight_pt[0][nh][-1], device=device, dtype=dtype)
+                        weight_npt[0][nh][-1] = torch.zeros_like(weight_npt[0][nh][-1], device=device, dtype=dtype)
+
+        if dtype == torch.half:
+            torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
+        else:
+            # High rtol seems necessary for
+            # test_native_multihead_attention_cpu_float32 on Windows,
+            # otherwise 2e-4 would likely be fine.
+            torch.testing.assert_close(ypt, ynpt, atol=2e-5, rtol=2e-3)
+
+        if need_weights:
+            torch.testing.assert_close(weight_pt, weight_npt)
+        else:
+            self.assertEqual(weight_pt, weight_npt)
+
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
+    def test_native_multihead_self_attention(self, device, dtype):
+        for (use_padding, pad_all) in ((False, False), (True, False), (True, True)):
+            for use_nt in (False, True):
+                # Figuring out exactly which elements of the weights are garbage in this
+                # case eludes me, and it's not particularly enlightening to test anyway
+                # because padding doesn't especially affect the intermediate weights.
+                for need_weights in (False, not pad_all):
+                    for average_attn_weights in (False, True):
+                        with self.subTest(use_padding=use_padding, pad_all=pad_all,
+                                          use_nt=use_nt, need_weights=need_weights,
+                                          average_attn_weights=average_attn_weights):
+                            self._test_multihead_attention_impl(
+                                device,
+                                dtype,
+                                "self",
+                                use_nt=use_nt,
+                                use_padding=use_padding,
+                                pad_all=pad_all,
+                                need_weights=need_weights,
+                                average_attn_weights=average_attn_weights,
+                            )
+
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
+    def test_native_multihead_encoder_decoder_attention(self, device, dtype):
```

Comment:
nit parametrize over the mode?

Reply:
the non-self modes aren't supported as well, so they aren't tested as thoroughly on purpose. (They are also currently unreachable.) This would be a good idea when it's time to test them all equally, though.